internal/socket.go

```diff
@@ -5,79 +5,33 @@ import (
 	"fmt"
 	"net"
 	"os/user"
+	"reflect"
 	"strconv"
 	"strings"
+	"sync"
 
 	"golang.org/x/sys/unix"
 	"google.golang.org/grpc/credentials"
 )
 
-// getUnixCreds returns info from unix socket connection about the process on the other end.
-func getUnixCreds(conn net.Conn) (*unix.Ucred, error) {
-	unixConn, ok := conn.(*net.UnixConn)
-	if !ok {
-		return nil, fmt.Errorf("socket is not a unix socket")
-	}
-
-	rawConn, err := unixConn.SyscallConn()
-	if err != nil {
-		return nil, fmt.Errorf("getting raw connection: %w", err)
-	}
-
-	var ucred *unix.Ucred
-	var internalErr error
-	err = rawConn.Control(func(fd uintptr) {
-		ucred, internalErr = unix.GetsockoptUcred(int(fd), unix.SOL_SOCKET, unix.SO_PEERCRED)
-	})
-	if internalErr != nil {
-		return nil, fmt.Errorf("doing GetsockoptUcred: %w", internalErr)
-	}
-	if err != nil {
-		return nil, fmt.Errorf("doing rawConn Control: %w", err)
-	}
-
-	if err := authenticateUser(ucred, []string{"nordvpn"}); err != nil {
-		return nil, err
-	}
-
-	return ucred, nil
-}
-
-func authenticateUser(ucred *unix.Ucred, allowGroups []string) error {
-	// root?
-	if ucred.Uid == 0 {
-		return nil
-	}
-	userInfo, err := user.LookupId(fmt.Sprintf("%d", ucred.Uid))
-	if err != nil {
-		return fmt.Errorf("authenticate user, lookup user info: %s", err)
-	}
-	// user belongs to the allowed group?
-	groups, err := userInfo.GroupIds()
-	if err != nil {
-		return fmt.Errorf("authenticate user, check user groups: %s", err)
-	}
-	for _, groupId := range groups {
-		groupInfo, err := user.LookupGroupId(groupId)
-		if err != nil {
-			return fmt.Errorf("authenticate user, check user group: %s", err)
-		}
-		for _, allowGroupName := range allowGroups {
-			if groupInfo.Name == allowGroupName {
-				return nil
-			}
-		}
-	}
-	return fmt.Errorf("requesting user does not have permissions")
+// NewUnixSocketCredentials create unix socket credentials structure
+func NewUnixSocketCredentials() *UnixSocketCredentials {
+	return &UnixSocketCredentials{}
 }
 
 // UnixSocketCredentials is used to retrieve linux user ID from unix socket connection between client and daemon
 // Implements credentials.TransportCredentials to be passed to gRPC server initialization
-type UnixSocketCredentials struct{}
+type UnixSocketCredentials struct {
+	mu sync.Mutex
+}
 
 // ServerHandshake is called when client connects to daemon.
 // We retrieve user ID which opened the client here.
-func (UnixSocketCredentials) ServerHandshake(c net.Conn) (net.Conn, credentials.AuthInfo, error) {
+func (cr *UnixSocketCredentials) ServerHandshake(c net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	// cgo calls to access user/group info are not thread safe, need to serialize
```

Comment:
The comment might not be correct, since the code works when not running in snap. 
The problem might be related to snap env instead of the c functions.